### PR TITLE
Feature/sdk improvements

### PR DESCRIPTION
**Pull request by dylanburkey**
## Overview
This PR introduces comprehensive improvements to the GAME SDK, focusing on enhanced testing, documentation, and core component reliability.

### 1. Core SDK Enhancements
- Enhanced agent implementation with improved state management
- Updated custom types and worker configurations for better type safety
- Improved worker implementation with better error handling
- Added new utilities module for common SDK operations
- Fixed API client to handle network errors and retries correctly

### 2. Test Suite Improvements
- Added comprehensive test suite with 94% code coverage
- Implemented network error handling tests with tenacity
- Added tests for:
  - API client and error handling
  - Agent functionality
  - Worker operations
  - Custom types
  - Configuration
  - Weather worker example

### 3. Documentation and Examples
- Added detailed API documentation:
  - Agent documentation
  - Worker documentation
  - Worker configuration guide
- Created getting started guide
- Added SDK overview
- Included example implementations:
  - Weather agent example
  - Weather worker example

### 4. Dependencies
- Added requirements.txt for better dependency management

## Testing
- All tests are passing with 94% code coverage
- Improved test reliability with better error handling
- Added test cases for edge cases and error conditions

## Documentation
Added comprehensive documentation to help developers:
- Understand SDK architecture
- Implement custom agents and workers
- Handle common use cases
- Follow best practices

## Breaking Changes
None. All changes are backward compatible.

## Next Steps
- Review and merge PR
- Update SDK version
- Release new documentation

### Code Coverage Testing

```

---------- coverage: platform darwin, python 3.9.6-final-0 -----------
Name                                 Stmts   Miss  Cover
--------------------------------------------------------
src/game_sdk/game/__init__.py            0      0   100%
src/game_sdk/game/agent.py              63     63     0%
src/game_sdk/game/api.py                31     31     0%
src/game_sdk/game/api_client.py         41     41     0%
src/game_sdk/game/api_v2.py             41     41     0%
src/game_sdk/game/config.py             13     13     0%
src/game_sdk/game/custom_types.py       63      1    98%
src/game_sdk/game/exceptions.py         15     15     0%
src/game_sdk/game/utils.py              66     66     0%
src/game_sdk/game/worker.py             16      0   100%
src/game_sdk/game/worker_config.py       8      0   100%
--------------------------------------------------------
TOTAL                                  357    271    24%


================================================== 16 passed in 0.38s ===================================================
```
```
---------- coverage: platform darwin, python 3.9.6-final-0 -----------
Name                                 Stmts   Miss  Cover
--------------------------------------------------------
src/game_sdk/game/__init__.py            0      0   100%
src/game_sdk/game/agent.py              63      9    86%
src/game_sdk/game/api.py                31     31     0%
src/game_sdk/game/api_client.py         41     41     0%
src/game_sdk/game/api_v2.py             41     41     0%
src/game_sdk/game/config.py             13     13     0%
src/game_sdk/game/custom_types.py       63      1    98%
src/game_sdk/game/exceptions.py         15      3    80%
src/game_sdk/game/utils.py              66     54    18%
src/game_sdk/game/worker.py             16      0   100%
src/game_sdk/game/worker_config.py      10      0   100%
--------------------------------------------------------
TOTAL                                  359    193    46%

============================================= 29 passed, 1 warning in 0.58s =============================================
```

## Updates
- Added the id field to the WorkerConfig class with a default value that generates a UUID.
- Fixed the ActionResponse initialization in test_agent_get_action test to properly serialize the response.
- Fixed the ```_get_action``` method in the Agent class to properly unpack the response dictionary when creating an ActionResponse object.
- The test coverage for the SDK main components is now:
```
worker.py: 100%
worker_config.py: 100%
custom_types.py: 98%
agent.py: 86%`
```

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform darwin, python 3.9.6-final-0 -----------
Name                                 Stmts   Miss  Cover
--------------------------------------------------------
src/game_sdk/game/__init__.py            0      0   100%
src/game_sdk/game/agent.py              63      9    86%
src/game_sdk/game/api.py                31      1    97%
src/game_sdk/game/api_client.py         41      0   100%
src/game_sdk/game/api_v2.py             41      0   100%
src/game_sdk/game/config.py             13      0   100%
src/game_sdk/game/custom_types.py       63      1    98%
src/game_sdk/game/exceptions.py         15      0   100%
src/game_sdk/game/utils.py              73     10    86%
src/game_sdk/game/worker.py             16      0   100%
src/game_sdk/game/worker_config.py      10      0   100%
--------------------------------------------------------
TOTAL                                  366     21    94%

============================================ 79 passed, 1 warning in 16.77s =============================================
```

## Updates:
- Added tenacity import to test_api_client.py to fix the network error handling tests.
- Updated the create_workers function in utils.py to:
- Use the correct worker configuration format for the API
- Properly handle worker functions and descriptions
- Use the correct endpoint (/v2/maps)
- Enhanced the validate_response function in utils.py to:
- Handle None responses
- Check for empty dictionaries
- Validate error status
- Check for None data values

The code coverage has also improved from 92% to 94%, indicating better test coverage of our codebase. The remaining untested code is primarily in the agent module (86% coverage), which we can address in future updates if needed.

### CODE REVIEW

File: examples/example_weather_agent.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        requests.RequestException: If the API call fails
+        
+    Example:
+        >>> result = get_weather("New York")
+        >>> print(result['message'])
+        'Current weather in New York: Sunny, 15°C, 60% humidity'
+    """
+    try:
+        # Simulate API call (replace with actual weather API in production)
+        response = requests.get('https://dylanburkey.com/assets/weather.json')
```
Comment (MantisClone) [inline examples/example_weather_agent.py:69]: Consider hosting weather.json on GitHub instead of on dylanburkey.com

File: examples/example_weather_agent.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+            clothing = 'Shorts and t-shirt'
+        elif temp > 15:
+            clothing = 'Light jacket'
+        else:
+            clothing = 'Sweater'
+            
+        return {
+            'status': 'success',
+            'message': f"Current weather in {city}: {weather_data['condition']}, "
+                      f"{weather_data['temp']}°F, {weather_data['humidity']} humidity",
```
Comment (MantisClone) [inline examples/example_weather_agent.py:91]: Please fix inconsistency. Weather data uses Celcius but the output message uses Fahrenheit.

File: examples/example_weather_agent.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    )
+    
+    # Create agent
+    agent = Agent(
+        api_key=api_key,
+        name="Weather Reporter",
+        agent_description="Reports weather and provides recommendations",
+        agent_goal="Help users prepare for weather conditions",
+        get_agent_state_fn=get_state,
+        workers=[worker_config]
```
Comment (MantisClone) [inline examples/example_weather_agent.py:195]: Oh nice. I didn't realize you set the workers list at initialization instead of adding them later.

File: examples/example_weather_agent.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        >>> agent = create_weather_agent()
+        >>> test_weather_agent(agent)
+        ✨ All weather reporter tests passed!
+    """
+    logger.debug("Starting weather reporter tests")
+    
+    # Test cities
+    test_cases = ["New York", "Miami", "Boston"]
+    
+    for city in test_cases:
```
Comment (MantisClone) [inline examples/example_weather_agent.py:225]: Consider using `@pytest.mark.parameterize` to specify test cases.

Comment (MantisClone) [reply examples/example_weather_agent.py:225]: Actually, please disregard my previous comments about using pytest. Maybe it's better to _not_ assume any given test framework.

File: examples/example_weather_worker.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            clothing = 'Shorts and t-shirt'
+        elif temp > 15:
+            clothing = 'Light jacket'
+        else:
+            clothing = 'Sweater'
+            
+        return {
+            'status': 'success',
+            'message': f"Current weather in {city}: {weather_data['condition']}, "
+                      f"{weather_data['temp']}°F, {weather_data['humidity']} humidity",
```
Comment (MantisClone) [inline examples/example_weather_worker.py:100]: Same consistency issue as the previous file - choose either Celcius or Fahrenheit.

File: examples/example_weather_worker.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        
+    except requests.RequestException as e:
+        logger.error(f"Failed to get weather: {e}")
+        return {
+            'status': 'error',
+            'message': f"Failed to get weather for {city}",
+            'data': None
+        }
+
+def get_worker_state(function_result: Optional[Dict], current_state: Optional[Dict]) -> Dict[str, Any]:
```
Comment (MantisClone) [inline examples/example_weather_worker.py:118]: I wonder if this should be named `get_and_update_worker_state()` to reflect that it doesn't just get the state, it also updates it.

**Review by MantisClone — COMMENTED** [review]
Looks great! It really helped me to understand the GAME python SDK!

This was only a partial review, covering the documentation but not the SDK code changes.
